File: scripts/fetch-huggingface-dataset.js
#!/usr/bin/env node
import { writeFileSync, existsSync } from "fs";
import { join } from "path";

async function convertToSqliteVec(dataset, outputDir) {
  const csvPath = join(outputDir, "synthetic_profiles.csv");

  if (existsSync(csvPath)) {
    console.log(`CSV file already exists at ${csvPath}, skipping conversion.`);
    return;
  }

  try {
    // Fetch dataset using the JSON API in batches
    const batchSize = 100; // Start smaller
    let offset = 0;
    let allRows = [];

    console.log("Fetching data from HuggingFace API...");

    while (true) {
      const viewerUrl = `https://datasets-server.huggingface.co/rows?dataset=${dataset}&config=default&split=train&offset=${offset}&length=${batchSize}`;
      console.log(`Fetching: ${viewerUrl}`);

      const response = await fetch(viewerUrl, {
        headers: {
          "User-Agent": "Mozilla/5.0 (compatible; NodeJS)",
        },
      });

      if (!response.ok) {
        const errorText = await response.text();
        console.log(
          `Failed to fetch batch at offset ${offset}: ${response.status} ${response.statusText}`,
        );
        console.log(`Error response: ${errorText}`);
        break;
      }

      const data = await response.json();

      if (!data.rows || data.rows.length === 0) {
        console.log(`No more rows found at offset ${offset}`);
        break;
      }

      allRows.push(...data.rows);
      console.log(`Fetched ${data.rows.length} rows (total: ${allRows.length})`);

      offset += batchSize;

      // Limit to prevent too much data - adjust as needed
      if (offset >= 5000) {
        console.log(`Limiting to first ${allRows.length} rows for demo purposes`);
        break;
      }
    }

    console.log(`Total rows fetched: ${allRows.length}`);

    // Generate CSV for easier parsing
    const csvHeaders = [
      "persona",
      "visit_id",
      "visit_time",
      "visit_description",
      "place_id",
      "url",
      "title",
      "domain",
      "visit_count",
      "interest",
      "title_name",
    ];

    const escapeCSV = (str) => {
      if (str === null || str === undefined) return "";
      const stringValue = String(str);
      // Escape quotes and wrap in quotes if contains comma, quote, or newline
      if (stringValue.includes(",") || stringValue.includes('"') || stringValue.includes("\n")) {
        return `"${stringValue.replace(/"/g, '""')}"`;
      }
      return stringValue;
    };

    let csv = csvHeaders.join(",") + "\n";

    for (const rowData of allRows) {
      const row = rowData.row;

      const csvRow = [
        escapeCSV(row.persona),
        escapeCSV(row.visit_id),
        escapeCSV(row.visit_time),
        escapeCSV(row.visit_description),
        escapeCSV(row.place_id),
        escapeCSV(row.url),
        escapeCSV(row.title),
        escapeCSV(row.domain),
        escapeCSV(row.visit_count),
        escapeCSV(row.interest),
        escapeCSV(row.title_name),
      ];

      csv += csvRow.join(",") + "\n";
    }

    const csvPath = join(outputDir, "synthetic_profiles.csv");
    writeFileSync(csvPath, csv);
    console.log(`Generated CSV file: ${csvPath} (${allRows.length} rows)`);
  } catch (error) {
    console.error("Error converting to SQLite Vec format:", error);
  }
}

async function fetchSyntheticProfiles() {
  try {
    console.log("Fetching synthetic profiles dataset from HuggingFace...");

    const dataset = "zijuncheng/synthetic_profiles_ver_0";

    // Create output directory in public so it can be accessed via HTTP
    const outputDir = "./public";

    // No need to download parquet file since we're using JSON API

    // Convert to SQLite Vec format
    console.log("Converting to SQLite Vec format...");
    await convertToSqliteVec(dataset, outputDir);

    console.log("Dataset download and conversion complete!");
  } catch (error) {
    console.error("Error fetching dataset:", error);
    process.exit(1);
  }
}

// Run the script
fetchSyntheticProfiles();

File: scripts/fetch_huggingface_dataset.py
#!/usr/bin/env python
"""
Fetch Hugging Face dataset and save to CSV using the datasets library.

Setup:
    1. Create a virtual environment:
       uv venv
    
    2. Install dependencies:
       uv pip install datasets pandas
    
    3. Run the script:
       uv run python scripts/fetch_huggingface_dataset.py
       
       Or with force refresh:
       uv run python scripts/fetch_huggingface_dataset.py --force

This will fetch the entire synthetic_profiles_ver_1 dataset (~58k rows)
from Hugging Face and save it to public/synthetic_profiles.csv
"""

from datasets import load_dataset
import pandas as pd
import sys

def fetch_dataset(force_refresh=False):
    print("Loading dataset from Hugging Face...")
    
    # Load the dataset (this will cache it locally)
    dataset = load_dataset("zijuncheng/synthetic_profiles_ver_1", split="train")
    
    print(f"Dataset loaded: {len(dataset)} rows")
    
    # Convert to pandas DataFrame
    df = pd.DataFrame(dataset)
    
    # Select only the columns we need for the CSV (matching the original format)
    columns_to_keep = [
        'persona', 'visit_id', 'visit_time', 'visit_description', 
        'place_id', 'url', 'title', 'domain', 'visit_count', 
        'interest', 'title_name'
    ]
    
    # Filter to only columns that exist
    available_columns = [col for col in columns_to_keep if col in df.columns]
    df = df[available_columns]
    
    # Save to CSV
    output_path = "public/synthetic_profiles.csv"
    df.to_csv(output_path, index=False)
    print(f"Saved {len(df)} rows to {output_path}")
    
    # Show unique personas
    unique_personas = df['persona'].unique()
    print(f"Unique personas found: {sorted(unique_personas)}")
    print(f"Persona distribution:")
    print(df['persona'].value_counts())

if __name__ == "__main__":
    force = "--force" in sys.argv or "-f" in sys.argv
    fetch_dataset(force)